system-addon/lib/TopSitesFeed.jsm

```diff
@@ -190,6 +197,33 @@ this.TopSitesFeed = class TopSitesFeed {
     }
   }
 
+  /**
+   * Set a custom screenshot for a topsite.
+   * @param url {string} The topsite URL to update with the new screenshot
+   * @param screenshotURL {string} The URL to use for the new screenshot
+   */
+  async setCustomTopsiteScreenshot({url, customScreenshotURL}) {
+    const pinnedLinks = await this.pinnedCache.request();
+    const frecentLinks = await this.frecentCache.request();
+    // Get the LinksCache entry that represents the topsite to update.
+    const link = pinnedLinks.concat(frecentLinks).find(topsite =>
+      topsite && topsite.url === url);
+
+    if (link) {
+      // If previous request failed, we need to clear any cached result
+      link.screenshotPreview = undefined;
+      Screenshots.maybeCacheScreenshot(link, customScreenshotURL, "screenshotPreview",
+        // When done, broadcast the result of the request.
+        screenshotPreview => {
+          this.store.dispatch(ac.BroadcastToContent({
```

Comment:
We probably don't need to broadcast to all pages when one form on a page is requesting the image? This would also change the main store's state which doesn't seem right for a transient image preview?

Reply:
@Mardak I agree with the comment but I'm not sure how to address this. Any advice on how to pass the message?

Reply:
I believe @sarracini was looking at adding something like `SendOnlyToContent` in https://bugzilla.mozilla.org/show_bug.cgi?id=1415491 but before then, I suppose the closest would be `SendToContent`, but that would still update main store.

@k88hudson is there a preferred way to send transient data to a single content page without updating main store?

Reply:
the Broadcast has remained the same atm.